example/FileSystem.tla

```diff
@@ -0,0 +1,138 @@
+------------------- MODULE FileSystem -----------------------
+\* A small example spec of some file system operations.
+\* This spec is meant for use in the `app.py` in the same directory,
+\* which demonstrates use of the chai client for interacting with
+\* the Apalache server.
+
+EXTENDS Apalache, Variants, Sequences, Integers, FiniteSets
+
+
+\*********\
+\* STATE *\
+\*********\
+
+CONSTANT
+    \* The names that can be used as path segments.
+    \* @type: Set(Str);
+    Names
+
+VARIABLES
+    \* The set of all files in the system.
+    \* @type: Set($file);
+    files,
+    \* The commend issued.
+    \* @type: $cmd;
+    cmd
+
+
+\**********\
+\* DOMAIN *\
+\**********\
+
+\** Representation of files, identified by their kind and their path.
+\*
+\* @typeAlias: path = Seq(Str);
+\* @typeAlias: file = File($path) | Dir($path);
+\* @type: $path => $file;
+File(path) == Variant("File", path)
+\* @type: $path => $file;
+Dir(path) == Variant("Dir", path)
+
+IsDir(f) == "Dir" = VariantTag(f)
+IsFile(f) == "File" = VariantTag(f)
+
+\* @type: $file => $path;
+FilePath(f) == VariantGetOrElse("Dir", f, VariantGetUnsafe("File", f))
+
+
+\** The children of a file (i.e., the contents of a directory)
+\*
+\* NOTE: The `Children` relation is reflexive: each file is its own child.
+\*
+\* @type: $file => Set($file);
+Children(p) ==
+  { child \in files: FilePath(p) = SubSeq(FilePath(child), 1, Len(FilePath(p))) }
```

Comment:
I think you want to take the `min` of `Len(FilePath(p))` and `Len(FilePath(child))` as 3rd argument to `SubSeq`:

```
If [...] n > Len(seq), then the result is undefined.
```
https://apalache.informal.systems/docs/lang/sequences.html?highlight=SubSeq#subsequence

Reply:
Since you're invoking `FilePath(_)` multiple times, you can wrap it in LETs. Also, you can avoid constructing a new sequence by doing a domain `\A` instead of SubSeq.
```tla
LET PathOfP == FilePath(p) IN
{ child \in files: 
  LET PathOfC == FilePath(child) IN
  /\  Len(PathOfC) >= Len(PathOfP)
  /\ \A i \in DOMAIN PathOfP: PathOfP[i] = PathOfC[i]
}
```

Reply:
> you can avoid constructing a new sequence by doing a domain \A instead of SubSeq

Are there expected performance improvements by doing this? On the face of it, it looks like just a more low-level approach to checking a subseq, like doing a manual for loop instead of taking a slice.

Reply:
Fixed in 749fecde29951e32923710c5eb0f83115cbeba49

Thanks for the offline explanation of the benefits here!